crates/fluvio-hub-util/src/fvm/api.rs

```diff
@@ -0,0 +1,116 @@
+//! Hub FVM API Client
+
+use anyhow::{Error, Result};
+use surf::get;
+use url::Url;
+
+use super::{Channel, PackageSet, RustTarget};
+
+/// Base URL for `fmv` related API requests
+pub const API_BASE_URL: &str = "https://hub.infinyon.cloud/hub/v1/fvm";
+
+/// HTTP Client for interacting with the Hub FVM API
+pub struct Client {
+    api_url: Url,
+}
+
+impl Default for Client {
+    fn default() -> Self {
+        // Safe to `unwrap`, `API_BASE_URL` is a valid URL and it's a constant
+        Self::new(Url::parse(API_BASE_URL).unwrap())
+    }
+}
+
+impl Client {
+    pub fn new(api_url: Url) -> Self {
+        Self { api_url }
+    }
+
+    /// Fetches a [`PackageSet`] from the Hub with the specific [`Channel`]
+    #[allow(unused)]
+    pub async fn fetch_package_set<S: AsRef<str>>(
+        &self,
+        name: S,
+        channel: &Channel,
+        arch: &RustTarget,
+    ) -> Result<PackageSet> {
+        let url = self.make_fetch_package_set_url(name, channel, arch)?;
+        let mut res = get(url).await.map_err(|err| {
+            tracing::error!("Failed to fetch package set: {}", err);
+            Error::msg(err.to_string())
+        })?;
+        let pkg = res.body_json::<PackageSet>().await.map_err(|err| {
+            tracing::error!("Failed to parse package set: {}", err);
+            Error::msg(err.to_string())
+        })?;
+
+        Ok(pkg)
+    }
+
+    /// Builds the URL to the Hub API for fetching a [`PackageSet`] using the
+    /// [`Client`]'s `api_url`.
+    fn make_fetch_package_set_url<S: AsRef<str>>(
```

Comment:
this seems like separate util?

Reply:
concern that this URL has too many nested so make it hard to maintain.
name, channel, arch seems to be qualifier.

Reply:
This builds the URL for the request separately so it can be isolated when testing, is strongly tied to the client given that is only used for this request.

Each of the params should be provided by the consumer:

- Name for the package set to fetch
- Version/Channel
- Architecture of the host machine to download compatible binaries